New profiler API

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48438 Output stacks (support for SVG visualization)
* **#48280 New profiler API**

Summary:
Adding new API for the kineto profiler that supports enable predicate
function - better support for profiling of training loops (example in test_profiler_kineto_api)

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

Commit:
New profiler API

Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
unit test

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
unit test

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
unit test

[ghstack-poisoned]

ilia-cher commented:
cc @gdankel for review too

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
unit test

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
unit test

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
unit test

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
unit test

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
unit test

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
unit test

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
unit test

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
unit test

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
unit test

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

guyang3532 commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,192 @@
+        elif act == EnablePred.Action.STOP_TRACE:
+            assert self.profiler is not None
+            self.profiler.__exit__(None, None, None)
+            assert self.enable_pred
```
This assertion will fail in case  self.enable_pred=None

ilia-cher replied:
right, good catch, thanks!

ilia-cher replied:
was going to add a non-enable_pred unit test too

ilia-cher replied:
fixed and added test

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

dzhulgakov commented:
@gchanan @mruberry @ngimel - for bike shedding on public api, especially `enable_pred` part

ilia-cher commented:
@gchanan @mruberry yes if you have thoughts on interface improvement (clearer/simpler) then let us know.

To make sure it is clear where the design comes from: typically we have warm and cold start up times coming from profiling libraries when profiling/tracing is enabled/re-enabled. That's fairly common and not only specific to CUPTI, but other libraries too, it's also, unfortunately, not always well documented.

enable_pred is optimized for the use case when profiler is added on top of the training loop, in that case we obviously don't want to profile the whole execution, but only some iterations. In this case it is also relatively 'easy' to avoid warmup issue since we naturally have a concept of iteration - we can do a few iterations with profiler enabled before starting recording of the trace.
That's exactly what enable_pred specifies - how many iterations wait/warmup/trace and a lambda to be called when profiling results are ready (to print/to save - see the test case for the example)
Another small improvement here is automatically marking the iterations in the trace with 'record_function' annotation.
When enable_pred is not used the profiler is identical to the old one in the behavior (doesn't care about warmup, traces everything).

mruberry commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+
+    In case output_fn is specified, it is called every time the trace is ready
+    """
+    class Action(object):
```
I think both Action and State can inherit from enum. See https://docs.python.org/3.5/library/enum.html.

ilia-cher replied:
yeah, i was curious if our minimum supported python version supports enums, also the same for dataclasses

mruberry commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+
+class EnablePred(object):
+    """
+    EnablePred describes on which steps profiler is active:
```
What does "EnablePred" stand for?

ilia-cher replied:
or ProfilerSchedule

mruberry commented:
> @gchanan @mruberry yes if you have thoughts on interface improvement (clearer/simpler) then let us know.

Happy to take a look. I was only able to give this a quick glance for now but the intent makes sense. I'll dedicate the time for a proper review tomorrow (Wednesday).

mruberry commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+        WARMUP = 1
+        ACTIVE = 2
+
+    def __init__(self, wait: int, warmup: int, active: int, output_fn: Optional[Callable[[prof.profile], None]]):
```
Suggest making the arguments to this function kwarg-only so it's more readable (it would also allow for default values)

ngimel commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+
+    Arguments:
+        activities - list of activity groups (CPU, CUDA)
+        enable_pred (optional) - iteration predicate function, used together with `next_step` call
```
can you describe what enable_pred is, and also give an example below of using profiler with `enable_pred`?

mruberry replied:
I'm not really sure what a reader is supposed to take away from this statement. What is an "iteration prediction function"? How is it used with `next_step`? Why is that relevant?

mruberry commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+    - profiler then starts actively tracing/collecting stats for the next 'active' steps
+    - after this, profiler returns to the inactive state and cycle repeats
+
+    In case output_fn is specified, it is called every time the trace is ready
```
This sounds like a callback instead of an "output_fn", is there a name involving "callback" that would fit?

mruberry replied:
Follow-up question, what is the signature of this function?

ngimel commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+    - profiler then starts actively tracing/collecting stats for the next 'active' steps
+    - after this, profiler returns to the inactive state and cycle repeats
+
+    In case output_fn is specified, it is called every time the trace is ready
```
an example of initialization will be useful here

ngimel commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+     - enable_pred is used for training loop tracing, allowing users to enable profiler on certain
+       iterations and account for the warmup
+     - when enable_pred is not set, profiler is always active
+     - next_step uses record_function api to add information about steps in the trace
```
1) The documentation does not explain what is `next_step` and does not explain what the user should do with it (the docs should say that users should call `next_step` once per training loop iteration, and that an event corresponding to the iteration will be recorded in this call)
2) There should be a link to record_function documentation here, otherwise record_function api is not clear

mruberry commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+
+class profile(object):
+    """
+    PyTorch profiler context manager.
```
Elaborate here on what the profiler can provide and how it's intended to be used. People should have an idea of what this looks like before reading the example. Mention the critical functions, how to use them, and why they're important.

mruberry replied:
Instead of mentioning the critical functions in this preamble, this could add docstrings to the methods.

ngimel commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+
+        if not self.enable_pred:
+            print("Warning: using profiler without enable predicate may result in the skewed " +
+                "results, use enable_pred to control the warmup time")
```
Give a more detailed error message, with an example of  how to instantiate and pass `enable_pred`

mruberry commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+        self.active = active
+        self.output_fn = output_fn
+
+        def active_active_fn(step):
```
The code here is very challenging to read and looks almost like a state machine. I guess I was expecting something simpler (like taking the modulus of the current iteration or epoch). Is this additional complexity really necessary?

ilia-cher replied:
it is indeed a state machine, yes

mruberry replied:
Can you elaborate on why that's necessary (vs taking a modulus)?

ilia-cher replied:
I can e.g. give an example showing that it's not so simple.
So you take the modulus and you know your current state, is it enough?
You still need to know what was the previous state and what where in the sequence you are.
If the current state is active, and the previous is warmup - action is start_trace
If the current state is active, and the previous is inactive - action is start_warmup, followed by start_trace
If the current state is active, and the previous state is active, does it mean we don't need to do anything?
It depends, consider (wait=0, warmup=0, active=2) - sometimes we don't do anything, sometimes we do stop, save, warmup, start.

So you still have to consider a lot of possibilities and edge cases, then in that case, why not just encode states and actions explicitly in the map? The original version of the code that didn't use it was way less readable imo.

mruberry replied:
What actually happens when wait=0, warmup=0 and active=2? I would expect the following pattern:

ACTIVE ACTIVE ACTIVE ACTIVE ACTIVE...

Is this not correct?

ilia-cher replied:
"wait=0, warmup=0 and active=2" means don't wait, don't do warm up, trace for two iterations

ilia-cher replied:
then the sequence repeats, and we provide access to the traces

mruberry replied:
Oh interesting. So it's more correctly:

ACTIVE ACTIVE REPORT ACTIVE ACTIVE REPORT

?

ilia-cher replied:
"REPORT" is implied at the every end of enable_pred cycle, will make sure to make it clear from the docs

mruberry commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+        self.step_rec_fn = prof.record_function("ProfilerStep#" + str(self.step_num))
+        self.step_rec_fn.__enter__()
+
+    def export_chrome_trace(self, path: str):
```
These will need docstrings

dzhulgakov commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+
+    def __enter__(self):
+        self.next_step()
+        if not self.enable_pred:
```
should you just have a default `enable_pred` which does this behavior?

ilia-cher replied:
in this case it's an endless tracing (for all iterations), then we should somehow represent it in EnablePred

dzhulgakov commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+        self.step_rec_fn: Optional[prof.record_function] = None
+
+        if not self.enable_pred:
+            print("Warning: using profiler without enable predicate may result in the skewed " +
```
prints are bad, use `warnings` module

dzhulgakov commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+        def inactive_warmup_fn(_):
+            raise RuntimeError("Incorrect profiler state sequence")
+
+        self.actions_map = {
```
I still feel it can be simplified. Can we replace the entire EnablePred with a function that returns one of the few enums of what to do on this iteration:
* NONE
* WARMUP (i.e. enable but not record)
* RECORD (i.e. enabled)
* RECORD_AND_FLUSH (i.e. enabled but dump a new iteration)

Then the profiler doesn't even need to depend on EnablePred class. It can be just an arbitrary callable with that maps `step_num` into `action`. Of course you'd provide the code you have here as the default implementation but the user can do their own. E.g. the simple one would be: `enable_pred = lambda num: WARMUP if num < 100 else RECORD` for simple profile that doesn't dump intermediate results.

This logic about state machine transition is best moved to Profiler directly

As for `output_fn` - what benefit does it have over just putting `if p.step % 100 == 0: <my code>` directly in the training loop? You can keep it, but I'd add it as a separate arg in addition to `enable_pred` so that both can be easily-passable lambdas

gdankel replied:
Good discussion here and above - I agree we should strive to simplify this as much as possible, and then make it as easy to understand as possible via good names and examples.

Warning: I haven't read the entire code yet so I might be missing the point - my comments are about user experience, not so much implementation details.

In my experience, a default behavior is be good for the majority of cases so I'll use that as example: one warmup iteration followed by three recorded iterations.
What frequently changes is when you want to trigger it (should allow simple modulo approach here I think) and what to do with the result. Less frequently, how many iterations to profile and which activities to include.

Given that, the standard use case ought to be something like:
trigger_when: p.step % 1000 == 100
handle_result: write_chrome_trace / print_summary / extract_metrics .... (list of predefined convenience functions)

i.e. trigger profiling at iteration 100, 1100 and so on. Important to note here is that profiling cannot be enabled while at the same time reporting. We need to warmup, record, then stop and process, in order for the overhead to be manageable during the record window.

gdankel replied:
To clarify the above, this will be the behavior:
1100: Warmup
1101-1102: Record
1103: Record, then stop and process, calling the handle_result function asynchronously when processing is complete. As much work as possible is deferred to the processing stage to minimize overhead during recording. Recording should not continue in parallel with processing for detailed profiling, but we also need to consider a light-weight continuous profiling use case, possibly streaming a trace or ongoing extraction of metrics.

ilia-cher replied:
@dzhulgakov  suggestion sounds good, basically let's provide a way to define a sequence of states explicitly through lambda + suggested default impl

one caveat - we might then allow users to specify sequences like: warmup none warmup none, etc - so we'll need to check for that, but i guess that's fine - assuming most users use supplied default and the power users know what they are doing;

@gdankel I think both solutions above should basically allow these kind of use cases

gdankel replied:
Sure... As long as users don't have to deal with this level of detail in the majority of cases, it seems reasonable.

dzhulgakov commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+            self.step_rec_fn.__exit__(None, None, None)
+        if self.profiler:
+            if self.enable_pred:
+                if self.enable_pred._num_state(self.step_num) == EnablePred.State.WARMUP:
```
calling state again here leaks details. Can't profile object remember its current state instead?

dzhulgakov commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+            if self.enable_pred:
+                if self.enable_pred._num_state(self.step_num) == EnablePred.State.WARMUP:
+                    self._run_action(EnablePred.Action.START_TRACE)
+            self._run_action(EnablePred.Action.STOP_TRACE, keep_profiler=True)
```
why is `keep_profiler` true in this case?

ilia-cher replied:
so that we can also use it after the context manager is finished

dzhulgakov commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,191 @@
+            enable_pred: Optional[EnablePred] = None,
+            record_shapes=False,
+            profile_memory=False,
+            with_stack=False):
```
I wonder whether we should still support `use_cuda` for backward compatibility?

dzhulgakov reviewed (changes requested):
Looks good, but I think it can be simplified a bit and other comments on docstrings need to be addressed

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function - better support for profiling of training loops (example in test_profiler_kineto_api)

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

ilia-cher commented:
Addressed comments about API and docs cc. @dzhulgakov @ngimel @gdankel

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function - better support for profiling of training loops (example in test_profiler_kineto_api)

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function - better support for profiling of training loops (example in test_profiler_kineto_api)

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

ngimel commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,283 @@
+
+    Arguments:
+        activities - list of activity groups (CPU, CUDA);
+        schedule - a callable takes step (int) as a single parameter and returns
```
Do we really want to allow people to pass callable here, instead of wait, warmup, active ints? Seems like writing a valid `schedule` would be hard and no one would do it?

ilia-cher replied:
will point to the user to use schedule()

ngimel commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,283 @@
+        schedule - a callable takes step (int) as a single parameter and returns
+          ProfilerAction value - specifies the profiler action on each step;
+        on_trace_ready (optional) - callable, called each time the trace is ready
+          during the profiling;
```
`trace is ready` is not defined. "when trace has been collected for the specified number of steps"?

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function - better support for profiling of training loops (example in test_profiler_kineto_api)

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function - better support for profiling of training loops (example in test_profiler_kineto_api)

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

ilia-cher commented:
website doc: 
[profiler.html.zip](https://github.com/pytorch/pytorch/files/5716179/profiler.html.zip)

Commit:
Update on "New profiler API"


Summary:
Adding new API for the kineto profiler that supports enable predicate
function - better support for profiling of training loops (example in test_profiler_kineto_api)

Test Plan:
python test/test_profiler.py -k test_profiler_kineto_api

Differential Revision: [D25142220](https://our.internmc.facebook.com/intern/diff/D25142220)

[ghstack-poisoned]

dzhulgakov commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,293 @@
+                warn("Incorrect schedule: RECORD followed by NONE")
+                self._stop_trace()
+            else:
+                assert prev_action == ProfilerAction.RECORD_AND_SAVE
```
you shouldn't use `assert` for user-induced errors. Change it for `raise` and better error message

dzhulgakov commented on torch/profiler/profiler.py:
```diff
@@ -0,0 +1,293 @@
+            elif prev_action == ProfilerAction.RECORD:
+                pass
+            else:
+                assert prev_action == ProfilerAction.RECORD_AND_SAVE
```
same here

dzhulgakov reviewed (commented):
Looks great, just the messages can be improved in one of the follow up PRs
